test/list_test.dart

```diff
@@ -724,4 +724,33 @@ Future<void> main([List<String>? args]) async {
     expect(team.players.length, 10);
     expect(realm.all<Person>().length, 100);
   });
+
+  test('ManagedRealmList.indexOf', () {
+    final config = Configuration.local([Team.schema, Person.schema]);
+    final realm = getRealm(config);
+
+    final team = Team('sad team', players: [for (int i = 0; i < 100; ++i) Person('$i')]);
+    realm.write(() => realm.add(team));
+    final players = team.players;
+
+    // expect(players, isA<ManagedRealmList<Person>>());
+    expect(players.indexOf(players.first, -1), 0); // okay to start from negative index
+    expect(players.indexOf(players.first, 1), -1); // start respected
+    expect(players.indexOf(players.first, 101), -1); // okay to start from non-existent index
+
+    var index = 0;
+    final r = Random(42); // deterministic
+    for (final p in players) {
+      expect(players.indexOf(p, r.nextInt(index + 1) - 1), index++);
+    }
+
+    // indexOf with wrong type of element, just returns -1. Proof:
+    final dartList = <int>[1, 2, 3];
+    expect((dartList as List<Object>).indexOf("abc"), -1); // ignore: unnecessary_cast
+
+    // Realm list behaves differently
+    expect(() => (players as List<Object>).indexOf(1), throwsA(isA<RealmException>())); // ignore: unnecessary_cast
+
+    expect(() => players.indexOf(Person('10')), throwsA(isA<RealmStateError>()));
```

Comment:
I find it weird that we expect this exception here but it is actually thrown indirectly in the `handle` getter above. I think we better move the throwing of this error to the more exact place. So probably in the `indexOf` method itself.

Reply:
I couldn't follow exectly where do we read the `handle` which will throw

Reply:
It is called in
```dart
RealmLinkHandle _getObjectAsLink(RealmObject object) {
  final realmLink = _realmLib.realm_object_as_link(object.handle._pointer);
  return RealmLinkHandle._(realmLink);
}
```

Reply:
This is a case of misuse by the user, we always `RealmStateError` in these cases, if we cannot prevent it compile time.

Reply:
I believe the pattern of calling `_handle!` is something we use in a lot of places. We can replace it with an assert, but I agree with Lubo that those preconditions are ideally checked at the API entrypoint as that's where we can provide most information. For example, this would allow us to throw a friendlier exception like `Can't call indexOf with an argument that is an unmanaged object`. You could argue that developers can potentially discover what went wrong by evaluating the stacktrace, but historically, we've had a lot of issues created where only the exception message has been available.

Reply:
So I can add the check earlier, but I still think it should be a `RealmStateError` since the user is mis-using the api in a way we cannot detect compile time. This is no different than say trying to listen for changes on a unmanaged or frozen list. In these cases we will also throw a `RealmStateError`

Reply:
Agree - not opposed to the exception type.

Reply:
Checked in `indexOf` checks if `element` is unmanaged, and throws more detailed `RealmStateError` now.